### PR TITLE
Generate thumbnail images for documents stored on disk / CouchDB

### PR DESCRIPTION
**Pull request by olegsv**
Generate preview thumbnails for images to improve loading speed.  

New Globals configuration parameters: 

1. Generate thumbnail – it's a boolean field, if it's false thumbnail feature won't work. (by default it's  false). 
2. Thumbnail size – maximum size of thumbnail both X and Y dimensions. (the generator calculate this by the image type (vertical or horizontal) ). 
 
Functionality. 

This generator uses PHP GD functions to make thumbnails. 

If Generate Thumbnails set to On, all uploaded images will be saved in 2 copies: an original file for download , and a thumbnail for presentation. 

There is option to present original image by adding additional parameter to reterive document url (http://localhost/main_openemr/openemr/controller.php?document&retrieve&patient_id=<PATIEN_ID>&document_id=<DOC_ID>) . 
New parameter - 'show_original=true' (in the end of all parameters). 

Limitations: 

1. Generate for images only, can't generate PDF / Word thumbnail yet.
2. This feature can't generate thumbnail when  encrypted file is uploaded.

Generator for existing files  

For backward compatibility we included a possibility to generate missing thumbnails. 

New button in Administration->Files category to start thumbnail generation for all images, this view returns all failed files.

### CODE REVIEW

**Comment by bradymiller** [comment]
Hi,
Recommend also posting a review request for this on the openemr developer forum to see if can get some other developers interested in reviewing this (I'm currently backlogged with the Innodb integration/testing and some other MU2 stuff). Do plan to look at it, but may be more than a week away at this point.
-brady

File: interface/patient_file/deleter.php [hunk 1]
```diff
@@ -141,14 +142,31 @@ function form_delete($formdir, $formid) {
...
+                @unlink(substr($thumb_url, 7));
+            }
+            break;
+        //for CouchDB store
+        // todo - make sure that missing deleting from CouchDB is bug.
+        /*case 1:
+            $couchDB = new CouchDB();
+            $couchDB->DeleteDoc($GLOBALS['couchdb_dbase'], $trow['couch_docid'], $trow['couch_revid']);
+            break;
+        */
```
Comment (bradymiller) [inline interface/patient_file/deleter.php]: This is odd that a delete doesn't remove the document in CouchDB.

File: interface/super/manage_site_files.php [hunk 2]
```diff
@@ -99,6 +100,42 @@
...
+ */
+
+switch((int)$GLOBALS['document_storage_method']){
+    case 0:
+        $storage_method = 'on Disc';
+        break;
+    case 1:
+        $storage_method = 'in CouchDB';
+        break;
+}
```
Comment (bradymiller) [inline interface/super/manage_site_files.php]: above block is not needed and not good from internationalization standpoint. Simply use the $GLOBALS['document_storage_method'] below to decide whether to state:
The thumbnails will be stored on Hard Disk
The thumbnails will be stored in CouchDB

File: interface/super/manage_site_files.php [hunk 3]
```diff
@@ -99,6 +100,42 @@
...
+
+if(isset($_POST['generate_thumbnails'])) {
+
+    $thumb_generator = new ThumbnailGenerator();
+    $results = $thumb_generator->generate_all();
+
+    $thumbnail_msg = "<p style='color: green'>" . xl('Generated thumbnail(s)') . " : {$results['sum_success']}</p>";
```
Comment (bradymiller) [inline interface/super/manage_site_files.php]: for new code, need to escape correctly:
``````
$thumbnail_msg = "<p style='color: green'>" . xtl('Generated thumbnail(s)') . " : " . text($results['sum_success']) . "</p>";

File: interface/super/manage_site_files.php [hunk 4]
```diff
@@ -99,6 +100,42 @@
...
+    $thumbnail_msg .= "<p style='color: red'>" . xl('Failed to generate') . ": {$results['sum_failed']}</p>";
```
Comment (bradymiller) [inline interface/super/manage_site_files.php]: for new code, need to escape correctly:
``````
$thumbnail_msg .= "<p style='color: red'>" . xlt('Failed to generate') . ": " . text($results['sum_failed']) . "</p>";

File: interface/super/manage_site_files.php [hunk 5]
```diff
@@ -99,6 +100,42 @@
...
+    foreach($results['failed'] as $key => $file){
+        $num = $key +1;
+        $thumbnail_msg .= "<p style='color: red; font-size: 11px'>$num.". " .$file</p>";
```
Comment (bradymiller) [inline interface/super/manage_site_files.php]: for new code, need to escape correctly:
``````
$thumbnail_msg .= "<p style='color: red; font-size: 11px'>" . text($num) . ". ."  . text($file) . "</p>";

File: interface/super/manage_site_files.php [hunk 6]
```diff
@@ -99,6 +100,42 @@
...
+    }
+} else {
+
+    $count_not_generated = ThumbnailGenerator::count_not_generated();
+
+    $thumbnail_msg = "<p>" .  xl('Files with empty thumbnail') . ": $count_not_generated</p>";
```
Comment (bradymiller) [inline interface/super/manage_site_files.php]: for new code, need to escape correctly:
``````
$thumbnail_msg = "<p>" .  xlt('Files with empty thumbnail') . ": " . text($count_not_generated) . "</p>";

File: interface/super/manage_site_files.php [hunk 7]
```diff
@@ -209,6 +259,28 @@ function msfFileChanged() {
 </center>
 
 </form>
+
+<div id="generate_thumb">
+    <table>
+        <tr>
+            <td class="thumb_title" width="19%">
+                <b><?php echo xl('Generate Thumbnails')?></b>
```
Comment (bradymiller) [inline interface/super/manage_site_files.php]: xlt()

File: interface/super/manage_site_files.php [hunk 8]
```diff
@@ -209,6 +259,28 @@ function msfFileChanged() {
...
+            </td>
+            <td class="thumb_subtitle" width="28%">
+                <span><?php echo xl('The thumbnails will be store ' . $storage_method) ?></span>
```
Comment (bradymiller) [inline interface/super/manage_site_files.php]: use strategy discussed above

File: interface/super/manage_site_files.php [hunk 9]
```diff
@@ -209,6 +259,28 @@ function msfFileChanged() {
...
+            </td>
+            <td class="thumb_msg" width="49%">
+                <span><?php echo $thumbnail_msg ?></span>
+            </td>
+            <td  class="thumb_form" width="16%" style="border-right:none">
+                <form method='post' action='manage_site_files.php#generate_thumb'>
+                    <input style="margin-top: 10px" type="submit" name="generate_thumbnails" value="<?php echo htmlspecialchars(xl('Generate')) ?>">
```
Comment (bradymiller) [inline interface/super/manage_site_files.php]: use this instead:
<input style="margin-top: 10px" type="submit" name="generate_thumbnails" value="<?php echo xla('Generate') ?>">

File: library/classes/CouchDB.class.php [hunk 10]
```diff
@@ -57,14 +57,35 @@ function createView($db){
...
+        $couch_json['pid'] = $patient_id;
+        $couch_json['encounter'] = $encounter;
+        $couch_json['mimetype'] = $type;
+        $couch_json['data'] = $json;
+        if($th_json) {
+            $couch_json['th_data'] = $th_json;
+        }
+
+        $resp = $this->send("PUT", "/".$db."/".$docid, json_encode($couch_json));
+        //$resp = $this->send("PUT", "/".$db."/".$docid, '{"_id":"'.$docid.'","pid":"'.$patient_id.'","encounter":"'.$encounter.'","mimetype":"'.$type.'","data":'.$json.'}');
```
Comment (bradymiller) [inline library/classes/CouchDB.class.php]: better to remove code than comment it out

File: library/classes/thumbnail/Thumbnail.class.php [hunk 11]
```diff
@@ -0,0 +1,178 @@
+<?php
+/**
+ * Make thumbnail file by PHP GD functions.
+ * version 1 - just for images.
+ *
+ * Another: Amiel Elboim
+ * Date: 29/05/16
+ */
```
Comment (bradymiller) [inline library/classes/thumbnail/Thumbnail.class.php]: Need a standard header. Here's an example:
http://www.open-emr.org/wiki/index.php/How_to_Document_Your_Code_Properly#Page-Level_DocBlock

File: sql/4_2_2-to-4_3_1_upgrade.sql [hunk 12]
```diff
@@ -71,4 +71,6 @@
 --  #IfNotListOccupation
 --    Custom function for creating Occupation List
 
-
+#IfMissingColumn documents thumb_url
+ALTER TABLE  `documents` ADD  `thumb_url` VARCHAR( 255 ) NULL DEFAULT NULL AFTER  `url`
```
Comment (bradymiller) [inline sql/4_2_2-to-4_3_1_upgrade.sql]: Why are there two NULL's in here?

Comment (bradymiller) [inline sql/4_2_2-to-4_3_1_upgrade.sql]: Remove the AFTER stuff.

**Comment by bradymiller** [comment]
Review is ongoing. Plan to test it out soon before complete the review.

File: interface/patient_file/deleter.php (hunk 1 above)
Comment (bradymiller) [inline interface/patient_file/deleter.php]: Agree with adding code to delete it in couchdb here.

File: library/classes/Document.class.php [hunk 13]
```diff
@@ -490,16 +503,36 @@ function change_patient($new_patient_id) {
...
    * @param  string  $mimetype     MIME type
    * @param  string  &$data        The actual data to store (not encoded)
+   * @param  string  $tmpfile      The tmp location of file
    * @param  string  $higher_level_path Optional subdirectory within the local document repository
    * @param  string  $path_depth   Number of directory levels in $higher_level_path, if specified
    * @param  integer $owner        Owner/user/service that is requesting this action
    * @return string                Empty string if success, otherwise error message text
    */
-  function createDocument($patient_id, $category_id, $filename, $mimetype, &$data,
+  function createDocument($patient_id, $category_id, $filename, $mimetype, &$data,$tmpfile,
```
Comment (bradymiller) [inline library/classes/Document.class.php]: Can't you instead use the mimetype, thus no need to bring in the $tmpfile??

File: controllers/C_Document.class.php [hunk 14]
```diff
@@ -399,27 +403,35 @@ function retrieve_action($patient_id="",$document_id,$as_file=true,$original_fil
...
-                else if ($disable_exit == "false") {
-                        $disable_exit=false;
-                }
-	    
+		if ($disable_exit == "true") {
+				$disable_exit=true;
+		}
+		else if ($disable_exit == "false") {
+				$disable_exit=false;
+		}
```
Comment (bradymiller) [inline controllers/C_Document.class.php]: In future, need to avoid bringing in whitespace changes as done above.

File: interface/patient_file/report/custom_report.php [hunk 15]
```diff
@@ -844,7 +844,7 @@ function next_prev(action){
                   else {
                     echo "<img src='" . $GLOBALS['webroot'] .
                       "/controller.php?document&retrieve&patient_id=&document_id=" .
-                      $document_id . "&as_file=false'><br><br>";
+                      $document_id . "&as_file=false&original_file=true&disable_exit=false&show_original=true'><br><br>";
```
Comment (bradymiller) [inline interface/patient_file/report/custom_report.php]: You should not need to set the following here since you are just setting what is already the default settings:
original_file
original_file
disable_exit

File: interface/patient_file/summary/demographics.php [hunk 16]
```diff
@@ -125,16 +125,17 @@ function image_widget($doc_id,$doc_catg)
...
         $image_file = $docobj->get_url_file();
+        $image_width = $GLOBALS['generate_thumb'] == 1 ? '' : 'width=100';
         $extension = substr($image_file, strrpos($image_file,"."));
         $viewable_types = array('.png','.jpg','.jpeg','.png','.bmp','.PNG','.JPG','.JPEG','.PNG','.BMP'); // image ext supported by fancybox viewer
         if ( in_array($extension,$viewable_types) ) { // extention matches list
                 $to_url = "<td> <a href = $web_root" .
-				"/controller.php?document&retrieve&patient_id=$pid&document_id=$doc_id" .
-				"/tmp$extension" .  // Force image type URL for fancybox
+				"/controller.php?document&retrieve&patient_id=$pid&document_id=$doc_id&as_file=false&original_file=true&disable_exit=false&show_original=true" .
+				"/tmp$extension" .  // Force image type URL for fancybo
```
Comment (bradymiller) [inline interface/patient_file/summary/demographics.php]: You do not need to set the following here since you are just setting what is already the default settings:
original_file
disable_exit

File: interface/super/manage_site_files.php (hunk 8 above)
Comment (bradymiller) [inline interface/super/manage_site_files.php]: Do note that each thumbnail needs to use whatever method the document uses for storage of the thumbnail. It does not matter what the globals is set to. For example, a installation may be using both hard drive and couchdb (if it changed from one to the other). So, when your are creating the thumbnails, you will use the setting for each document when you decide whether to use hard drive or couchdb.

File: library/classes/thumbnail/Thumbnail.class.php [hunk 17]
```diff
@@ -0,0 +1,178 @@
...
+        if(!is_null($max_size))$this->max_size = $max_size;
+
+    }
+
+    /**
+     * Check if system could make thumbnail for current file.
+     * @param (string) path to file
+     * @return (boolean)
+     */
+    public function file_support_thumbnail($file){
```
Comment (bradymiller) [inline library/classes/thumbnail/Thumbnail.class.php]: Wouldn't it be better to check the mimetype rather than the file?

File: library/classes/thumbnail/ThumbnailGenerator.php [hunk 18]
```diff
@@ -0,0 +1,184 @@
...
+        }
+
+        $resource = $this->thumb_obj->create_thumbnail($url);
+        if(!$resource) {
+            return false;
+        }
+
+        $thumb_name = $this->get_thumb_name($path_parts['basename']);
+        $full_path = $path_parts['dirname'] . '/' . $thumb_name;
+        $new_thumb_file = $this->thumb_obj->image_to_file($resource, $full_path);
```
Comment (bradymiller) [inline library/classes/thumbnail/ThumbnailGenerator.php]: It doesn't look like you are correctly figuring out the path. See the algorithm in the documents class for displaying and saving files to ensure done correctly (note it's not simply a matter of copying the url or else it would break if changed the path of openemr).

**Comment by bradymiller** [comment]
Review is complete. I am assuming you have tested your code on couchdb. Looking forward to the next revision.

File: library/classes/thumbnail/Thumbnail.class.php (hunk 17 above)
Comment (matrix-amiel) [inline library/classes/thumbnail/Thumbnail.class.php]: What with encrypted files? Those with good image extension but they will fail just in this checking..

File: interface/patient_file/summary/demographics.php (hunk 16 above)
Comment (matrix-amiel) [inline interface/patient_file/summary/demographics.php]: I tried it but if I don’t send all the variables in the url ‘Document class’ put true in the ‘original_file’ (first variable) instead ‘show_original’, current logic doesn’t check a variable name..

File: interface/patient_file/report/custom_report.php (hunk 15 above)
Comment (matrix-amiel) [inline interface/patient_file/report/custom_report.php]: I tried it but if I don’t send all the variables in the url ‘Document class’ put true in the ‘original_file’ (first variable) instead ‘show_original’, current logic doesn’t check a variable name..

File: interface/patient_file/summary/demographics.php (hunk 16 above)
Comment (bradymiller) [inline interface/patient_file/summary/demographics.php]: Not surprised(the controller also mangles booleans). It's not pretty, but then would keep what you have.

File: interface/patient_file/report/custom_report.php (hunk 15 above)
Comment (bradymiller) [inline interface/patient_file/report/custom_report.php]: Not surprised(the controller also mangles booleans). It's not pretty, but then would keep what you have.

**Comment by olegsv** [comment]
@bradymiller we have submitted changes. can you check again ?

**Comment by bradymiller** [comment]
Next in my queue :)
http://www.bradymd.com/todo/#openemr-code-reviews

File: library/classes/Document.class.php (hunk 13 above)
Comment (bradymiller) [inline library/classes/Document.class.php]: Two things here:
1. Can't use mimetype rather than bringing in $tmpfile.
2. Placing a parameter in the middle seems very dangerous. Since this function is used in other places in the codebase, wouldn't it be safer to add the parameter at the end?

File: library/classes/thumbnail/Thumbnail.class.php [hunk 19]
```diff
@@ -0,0 +1,192 @@
...
+ * @author  Amiel Elboim <amielel@matrix.co.il>
+ * @link    http://www.open-emr.org
+ */
+
+class Thumbnail
+{
+
+    //max size of thumbnail
+    public $max_size = 100;
+    public $thumbnail_type = 'jpg';
```
Comment (bradymiller) [inline library/classes/thumbnail/Thumbnail.class.php]: I am confused here. Is the thumbnail type hard-coded to jpg always?

File: library/classes/thumbnail/Thumbnail.class.php [hunk 20]
```diff
@@ -0,0 +1,192 @@
...
+ * @package OpenEMR
+ * @author  Amiel Elboim <amielel@matrix.co.il>
+ * @link    http://www.open-emr.org
+ */
+
+class Thumbnail
+{
+
+    //max size of thumbnail
+    public $max_size = 100;
```
Comment (bradymiller) [inline library/classes/thumbnail/Thumbnail.class.php]: Should this be using the thumb_size global?

File: library/globals.inc.php [hunk 21]
```diff
@@ -1115,7 +1115,18 @@
       'Lab Report',                     // default
       xl('Document category name for storage of electronically received MDM documents.')
     ),
-
+    'generate_thumb' => array(
```
Comment (bradymiller) [inline library/globals.inc.php]: be more specific:
generate_doc_thumb

File: library/globals.inc.php [hunk 22]
```diff
@@ -1115,7 +1115,18 @@
...
+        xl('Generate thumbnail'),
+        'bool',
+        '0',
+        xl('Generate thumbnail images'),
+    ),
+    'thumb_size' => array(
```
Comment (bradymiller) [inline library/globals.inc.php]: be more specific:
thumb_doc_max_size

**Comment by bradymiller** [comment]
Completed another review . This is getting close. Guessing it will be ready on your next revision (if address my review comments). I am assuming that you have tested this on CouchDB.

File: library/classes/Document.class.php [hunk 23]
```diff
@@ -490,16 +503,36 @@ function change_patient($new_patient_id) {
...
     $higher_level_path='', $path_depth=1, $owner=0) {
     // The original code used the encounter ID but never set it to anything.
```
Comment (bradymiller) [inline library/classes/Document.class.php]: Two things here:
1. Can't use mimetype rather than bringing in $tmpfile(I am just looking for reasoning why doing this; if there is a reason, then would be ok).
2. Placing a new parameter in the middle seems very dangerous. Since this function is used in other places in the codebase, wouldn't it be much safer to add the parameter at the end?

File: library/classes/thumbnail/Thumbnail.class.php (hunk 19 above)
Comment (olegsv) [inline library/classes/thumbnail/Thumbnail.class.php]: Yes. Thumbnail type does not depend on content type. In future we can add configuration gif/png/jpg if we choose to.

Comment (bradymiller) [inline library/classes/thumbnail/Thumbnail.class.php]: Ah, makes sense.

**Comment by matrix-amiel** [comment]
Yes. of course, I checked it on couchDB, working well.

File: sites/default/sqlconf.php [hunk 24]
```diff
@@ -1,16 +1,17 @@
 <?php
```
Comment (matrix-amiel) [inline sites/default/sqlconf.php]: oooops.. sorry. must remove this.

File: library/classes/Document.class.php [hunk 25]
```diff
@@ -587,6 +631,14 @@ function createDocument($patient_id, $category_id, $filename, $mimetype, &$data,
       if (file_put_contents($filepath . $filename, $data) === FALSE) {
         return xl('Failed to create') . " $filepath$filename";
       }
+	  if( $has_thumbnail ) {
+		 $this->thumb_url = "file://" . $filepath . $this->get_thumb_name($filename);
+		 // Store the file into its proper directory.
+		 if (file_put_contents($filepath . 'th_' . $filename, $thumbnail_data) === FALSE) {
```
Comment (bradymiller) [inline library/classes/Document.class.php]: Shouldn't this use the get_thumb_name() function rather than adding the th_ to ensure is the same?

File: library/classes/thumbnail/ThumbnailGenerator.php [hunk 26]
```diff
@@ -0,0 +1,207 @@
+    <?php
+/**
+ * Adding thumbnails to all files
+ * User: amiel
+ * Date: 30/05/16
+ * Time: 14:55
+ */
```
Comment (bradymiller) [inline library/classes/thumbnail/ThumbnailGenerator.php]: Needs a standard gpl header like you placed in above library/classes/thumbnail/Thumbnail.class.php file

Comment (bradymiller) [inline library/classes/thumbnail/ThumbnailGenerator.php]: I fixed this in the code brought into the main codebase.

File: library/classes/Document.class.php (hunk 25 above)
Comment (bradymiller) [inline library/classes/Document.class.php]: Note I did not address this issue in the code I brought into the codebase. Please submit a separate commit to fix this issue.

**Comment by bradymiller** [comment]
Just committed this to the main codebase. Thanks for the nice feature.
